crates/agglayer-certificate-orchestrator/src/certificate_task.rs

```diff
@@ -0,0 +1,282 @@
+use std::sync::Arc;
+
+use agglayer_storage::stores::{StateReader, StateWriter};
+use agglayer_types::{Certificate, CertificateHeader, CertificateStatus, CertificateStatusError};
+use tokio::sync::{mpsc, oneshot};
+use tracing::{debug, error, trace, warn};
+
+use crate::{network_task::NetworkTaskMessage, Certifier, EpochPacker};
+
+/// A task that processes a certificate, including certifying it and settling
+/// it.
+///
+/// Once the `process` function is called, this task will handle everything
+/// related to the certificate until it gets finalized, including exchanging the
+/// required messages with the network task to both get required information
+/// from it and notify it of certificate progress.
+pub struct CertificateTask<StateStore, CertifierClient, SettlementClient> {
+    certificate: Certificate,
+    header: CertificateHeader,
+
+    network_task: mpsc::Sender<NetworkTaskMessage>,
+    state_store: Arc<StateStore>,
+    certifier_client: Arc<CertifierClient>,
+    settlement_client: Arc<SettlementClient>,
+}
+
+impl<StateStore, CertifierClient, SettlementClient>
+    CertificateTask<StateStore, CertifierClient, SettlementClient>
+where
+    StateStore: StateReader + StateWriter,
+    CertifierClient: Certifier,
+    SettlementClient: EpochPacker,
+{
+    pub fn new(
+        certificate: Certificate,
+        header: CertificateHeader,
```

Comment:
```suggestion
        certificate: Certificate,
        header: CertificateHeader,
```

This method could encapsulate the db call to fetch the `CertificateHeader`, to avoid the fact that we could initiate a `CertificateTask` with a header that is different from the certificate passed.